controllers/openstackmachine_controller.go

```diff
@@ -690,6 +698,141 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 	return ctrl.Result{}, nil
 }
 
+func (r *OpenStackMachineReconciler) getMachineServer(ctx context.Context, openStackMachine *infrav1.OpenStackMachine) (*infrav1alpha1.OpenStackServer, error) {
+	machineServer := &infrav1alpha1.OpenStackServer{}
+	machineServerName := client.ObjectKey{
+		Namespace: openStackMachine.Namespace,
+		Name:      openStackMachine.Name,
+	}
+	err := r.Client.Get(ctx, machineServerName, machineServer)
+	if err != nil {
+		return nil, err
+	}
+	return machineServer, nil
+}
+
+func openStackMachineSpecToOpenStackServerSpec(openStackMachineSpec *infrav1.OpenStackMachineSpec, resolved *infrav1.ResolvedMachineSpec, identityRef infrav1.OpenStackIdentityReference, tags []string, failureDomain string, userData string) (*infrav1alpha1.OpenStackServerSpec, error) {
+	if resolved == nil {
+		return nil, errors.New("resolved machine spec is nil")
+	}
+
+	openStackServerSpec := &infrav1alpha1.OpenStackServerSpec{
+		AdditionalBlockDevices: openStackMachineSpec.AdditionalBlockDevices,
+		ConfigDrive:            openStackMachineSpec.ConfigDrive,
+		Flavor:                 openStackMachineSpec.Flavor,
+		IdentityRef:            identityRef,
+		Image:                  infrav1.ImageParam{ID: &resolved.ImageID},
+		RootVolume:             openStackMachineSpec.RootVolume,
+		ServerMetadata:         openStackMachineSpec.ServerMetadata,
+		SSHKeyName:             openStackMachineSpec.SSHKeyName,
+		Tags:                   tags,
+	}
+	if resolved.ServerGroupID != "" {
+		openStackServerSpec.ServerGroup = &infrav1.ServerGroupParam{ID: &resolved.ServerGroupID}
+	}
+
+	if failureDomain != "" {
+		openStackServerSpec.AvailabilityZone = &failureDomain
+	}
+
+	if userData != "" {
+		openStackServerSpec.UserData = &userData
+	}
+
+	if openStackMachineSpec.Trunk {
+		openStackServerSpec.Trunk = ptr.To(true)
+	}
+
+	serverPorts := make([]infrav1.PortOpts, len(resolved.Ports))
+	for i, port := range resolved.Ports {
+		serverPorts[i] = infrav1.PortOpts{}
+		if port.NetworkID == "" {
+			return nil, errors.New("port network ID is empty")
+		}
+		serverPorts[i].Network = &infrav1.NetworkParam{
+			ID: ptr.To(port.NetworkID),
+		}
+		if port.Description != "" {
+			serverPorts[i].Description = ptr.To(port.Description)
+		}
+		if len(port.FixedIPs) > 0 {
+			serverPorts[i].FixedIPs = make([]infrav1.FixedIP, len(port.FixedIPs))
+			if port.FixedIPs[0].SubnetID != nil {
+				serverPorts[i].FixedIPs = []infrav1.FixedIP{
+					{
+						Subnet: &infrav1.SubnetParam{
+							ID: port.FixedIPs[0].SubnetID,
+						},
+					},
+				}
+			}
+			if port.FixedIPs[0].IPAddress != nil {
```

Comment:
also why only one FixedIP is used? would dual-stack support be affected?

Reply:
that's a miss as well. I'll fix it with a loop.